### PR TITLE
File serialization/deserialization support

### PR DESCRIPTION
**Pull request by 2Dsharp**
Documents can be saved and read from the custom FXD format with a magic number sequence to validate the file integrity (needs to be improved). 

Improved file open/import system with a cache implementation for currently opened subwindows.
Some memleaks were fixed/avoided and raw pointers switched to smart pointers. 
And added minor refactor and clean up.

### CODE REVIEW

File: src/tools/components/boundingrectitem.cpp [hunk 1]
```diff
@@ -23,7 +23,7 @@ void BoundingRectItem::paint(QPainter *painter,
                              const QStyleOptionGraphicsItem *option,
```
Comment (hellozee) [inline src/tools/components/boundingrectitem.cpp]: Remove this file if possible, will be overwritten by my commits anyway

File: src/tools/components/boundingrectitem.h [hunk 2]
```diff
@@ -29,6 +29,7 @@ class BoundingRectItem : public QGraphicsRectItem
```
Comment (hellozee) [inline src/tools/components/boundingrectitem.h]: This file too

File: src/mainwindow.cpp [hunk 3]
```diff
@@ -238,22 +322,89 @@ void MainWindow::on_actionExit_triggered()
...
+    QSharedDataPointer<Canvas> canvas = paintWidget->getCanvas();
+    Document document(layers, canvas);
+
+    storeDocument(fileName, document);
+    updateStateChange(SAVED, fileName, currentWindow);
+}
+
+QString MainWindow::getFileName(QString &fileName)
+{
+    if (!fileName.endsWith(".fxt") && !fileName.endsWith(".fxd")) {
```
[the remark this answers is not in the document]
Comment (2Dsharp) [reply src/mainwindow.cpp]: Just for the sake of being natural to the mind. Someone might mistake `fxdee` to `fxtee`.

File: src/model/document.cpp [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    QList<Layer*> layers;
+    Canvas *canvas = new Canvas(_name);
+
+    qint64 magicNum;
+    qint32 version;
+
+    in >> magicNum >> version;
+
+    if (magicNum != _magicNum) {
+        //emit badFileFormat("The file selected is either corrupted or not selected");
```
[the remark this answers is not in the document]
Comment (2Dsharp) [reply src/model/document.cpp]: Yes. We should really be throwing exceptions in here, task for the next iteration.

File: src/model/document.cpp [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        return;
+    }
+
+    in  >> *canvas >> layers;
+
+    _layers = layers;
+    _canvas = QSharedDataPointer<Canvas>(canvas);
+
+    // To check if canvas is working properly
+    //qDebug() << _canvas->getResolution() << _canvas->getWidth();
```
[the remark this answers is not in the document]
Comment (2Dsharp) [reply src/model/document.cpp]: Done.

File: src/tools/components/boundingrectitem.cpp (hunk 1 above)
Comment (2Dsharp) [reply src/tools/components/boundingrectitem.cpp]: Some dependencies might break on this version, let's remove them while merging your branch instead.

Comment (hellozee) [reply src/tools/components/boundingrectitem.cpp]: fine, merge the branch then

**Comment by 2Dsharp** [comment]
Getting it merged.

**Pull request merged**